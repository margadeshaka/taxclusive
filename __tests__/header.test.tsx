import { render, screen, fireEvent } from "@testing-library/react";
import Header from "@/components/header";
import { usePathname } from "next/navigation";
import renderer from "react-test-renderer";

// Mock the Next.js navigation hooks
jest.mock("next/navigation", () => ({
  usePathname: jest.fn(),
}));

describe("Header", () => {
  beforeEach(() => {
    // Reset mocks before each test
    jest.clearAllMocks();

    // Mock the usePathname hook to return '/' by default
    (usePathname as jest.Mock).mockReturnValue("/");

    // Mock window.scrollY
    Object.defineProperty(window, "scrollY", {
      configurable: true,
      value: 0,
    });
  });

  it("renders the logo and navigation links", () => {
    render(<Header />);

    // Check if the logo is rendered
    expect(screen.getByText("Taxclusive")).toBeInTheDocument();

    // Check if navigation links are rendered
    expect(screen.getByRole("link", { name: "Home" })).toBeInTheDocument();
    expect(screen.getByRole("link", { name: "About Us" })).toBeInTheDocument();
    expect(screen.getByRole("link", { name: "Services" })).toBeInTheDocument();
    expect(screen.getByRole("link", { name: "Expertise" })).toBeInTheDocument();
    expect(screen.getByRole("link", { name: "Insights" })).toBeInTheDocument();
    expect(screen.getByRole("link", { name: "Blogs" })).toBeInTheDocument();
    expect(screen.getByRole("link", { name: "FAQ" })).toBeInTheDocument();
    expect(screen.getByRole("link", { name: "Contact" })).toBeInTheDocument();

    // Check if CTA buttons are rendered
    expect(screen.getByRole("link", { name: "Book an Appointment" })).toBeInTheDocument();
    expect(screen.getByRole("link", { name: "Get in Touch" })).toBeInTheDocument();
  });

  it("highlights the active navigation link", () => {
    // Mock the usePathname hook to return '/about'
    (usePathname as jest.Mock).mockReturnValue("/about");

    render(<Header />);

    // Get all navigation links
    const homeLink = screen.getByRole("link", { name: "Home" });
    const aboutLink = screen.getByRole("link", { name: "About Us" });

    // Check if the active link has the correct class
    expect(homeLink.className).not.toContain("text-primary font-bold");
    expect(aboutLink.className).toContain("text-primary font-bold");
  });

  it("toggles the mobile menu when the menu button is clicked", () => {
    render(<Header />);

    // Get the menu button
    const menuButton = screen.getByRole("button", { name: "Open menu" });

    // Mobile menu should not be visible initially
    expect(screen.queryByRole("navigation", { name: "Mobile" })).not.toBeInTheDocument();

    // Click the menu button to open the menu
    fireEvent.click(menuButton);

    // Mobile menu should now be visible
    expect(screen.getByRole("navigation", { name: "Mobile" })).toBeInTheDocument();

    // Menu button should now be labeled "Close menu"
    expect(screen.getByRole("button", { name: "Close menu" })).toBeInTheDocument();

    // Click the menu button again to close the menu
    fireEvent.click(screen.getByRole("button", { name: "Close menu" }));

    // Mobile menu should no longer be visible
    expect(screen.queryByRole("navigation", { name: "Mobile" })).not.toBeInTheDocument();
  });

  it("changes header appearance on scroll", () => {
    render(<Header />);

    // Get the header element
    const header = screen.getByRole("banner");

    // Initially, header should not have the scrolled class
    expect(header.className).not.toContain("backdrop-blur");

    // Simulate scrolling
    Object.defineProperty(window, "scrollY", {
      configurable: true,
      value: 20,
    });

    // Trigger the scroll event
    fireEvent.scroll(window);

    // Header should now have the scrolled class
    expect(header.className).toContain("backdrop-blur");
  });

  it("matches snapshot", () => {
    // Mock the usePathname hook to return '/' for consistent snapshots
    (usePathname as jest.Mock).mockReturnValue("/");

    const tree = renderer.create(<Header />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
