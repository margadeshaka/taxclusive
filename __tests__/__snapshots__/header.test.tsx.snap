// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header matches snapshot 1`] = `
<header
  aria-label="Main navigation"
  className="sticky top-0 z-50 w-full border-b transition-all duration-300 bg-background"
>
  <div
    className="container flex h-16 items-center justify-between"
  >
    <div
      className="flex items-center gap-2"
    >
      <a
        aria-label="Taxclusive - Home"
        className="flex items-center gap-2 font-serif font-bold text-xl pt-4"
        href="/"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onTouchStart={[Function]}
      >
        <img
          alt="Taxclusive Logo"
          data-nimg="1"
          decoding="async"
          height={100}
          loading="lazy"
          onError={[Function]}
          onLoad={[Function]}
          src="/_next/image?url=%2Flogo.png&w=640&q=75"
          srcSet="/_next/image?url=%2Flogo.png&w=256&q=75 1x, /_next/image?url=%2Flogo.png&w=640&q=75 2x"
          style={
            {
              "color": "transparent",
            }
          }
          width={250}
        />
      </a>
    </div>
    <nav
      aria-label="Main"
      className="hidden md:flex gap-6"
    >
      <a
        className="text-sm font-medium transition-colors hover:text-primary text-primary font-bold"
        href="/"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onTouchStart={[Function]}
      >
        Home
      </a>
      <a
        className="text-sm font-medium transition-colors hover:text-primary "
        href="/about"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onTouchStart={[Function]}
      >
        About
      </a>
      <a
        className="text-sm font-medium transition-colors hover:text-primary "
        href="/services"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onTouchStart={[Function]}
      >
        Services
      </a>
      <a
        className="text-sm font-medium transition-colors hover:text-primary "
        href="/expertise"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onTouchStart={[Function]}
      >
        Expertise
      </a>
      <a
        className="text-sm font-medium transition-colors hover:text-primary "
        href="/blogs"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onTouchStart={[Function]}
      >
        Insights
      </a>
      <a
        className="text-sm font-medium transition-colors hover:text-primary "
        href="/faq"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onTouchStart={[Function]}
      >
        FAQ
      </a>
      <a
        className="text-sm font-medium transition-colors hover:text-primary "
        href="/contact"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onTouchStart={[Function]}
      >
        Contact
      </a>
    </nav>
    <div
      className="hidden md:flex items-center gap-4"
    >
      <a
        className="inline-flex h-10 items-center justify-center rounded-md px-6 text-sm font-medium shadow transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring bg-primary text-primary-foreground hover:bg-primary/90"
        href="/contact"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onTouchStart={[Function]}
      >
        Book Consultation
      </a>
      <a
        className="inline-flex h-10 items-center justify-center rounded-md bg-primary px-6 text-sm font-medium text-primary-foreground shadow transition-colors hover:bg-primary/90 focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring"
        href="/contact"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onTouchStart={[Function]}
      >
        Get in Touch
      </a>
    </div>
    <button
      aria-controls="mobile-menu"
      aria-expanded={false}
      aria-label="Open menu"
      className="flex items-center md:hidden"
      onClick={[Function]}
    >
      <svg
        className="lucide lucide-menu h-6 w-6"
        fill="none"
        height={24}
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth={2}
        viewBox="0 0 24 24"
        width={24}
        xmlns="http://www.w3.org/2000/svg"
      >
        <line
          x1="4"
          x2="20"
          y1="12"
          y2="12"
        />
        <line
          x1="4"
          x2="20"
          y1="6"
          y2="6"
        />
        <line
          x1="4"
          x2="20"
          y1="18"
          y2="18"
        />
      </svg>
    </button>
  </div>
</header>
`;
